### ISSUE TITLE
ICE: Cyclic trait resolution with associated types in nested impls

### ISSUE THREAD

**Issue by trait-Cong** [rust-lang/rust#150954]
I tried this code:

```rust
trait Component<M> {
    type Interface;
}
trait HasComponent<I> {}
struct Foo;
impl HasComponent<<Foo as Component<Foo>>::Interface> for Foo
where
    Foo: Component,
{}
impl<M: HasComponent<()>> Component<M> for Foo {
    type Interface = u8;
}
fn main() {}

```

I expected to see this happen: A clear compilation error stating that a cyclic dependency was detected or that the recursion limit was reached (e.g., E0391 or standard overflow error)

Instead, this happened: The compiler panicked with an assertion failure in `rustc_trait_selection.`

### Meta

`rustc --version --verbose`:
```
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

<details><summary>Backtrace</summary>
<p>

```
error[E0107]: missing generics for trait `Component`
 --> ../../try_pro/bug6.rs:8:10
  |
8 |     Foo: Component,
  |          ^^^^^^^^^ expected 1 generic argument
  |
note: trait defined here, with 1 generic parameter: `M`
 --> ../../try_pro/bug6.rs:1:7
  |
1 | trait Component<M> {
  |       ^^^^^^^^^ -
help: add missing generic argument
  |
8 |     Foo: Component<M>,
  |                   +++

error: internal compiler error: compiler/rustc_trait_selection/src/traits/select/mod.rs:2496:17: impl DefId(0:10 ~ bug6[5ceb]::{impl#0}) was matchable against Binder { value: TraitPredicate(<Foo as HasComponent<()>>, polarity:Positive), bound_vars: [] } but now is not


thread 'rustc' (2619225) panicked at compiler/rustc_trait_selection/src/traits/select/mod.rs:2496:17:
Box<dyn Any>
stack backtrace:
   0:     0x7fed21c33ab3 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h718e2d17a1928e63
   1:     0x7fed22401b98 - core::fmt::write::h1d2246b072ea91eb
   2:     0x7fed239cd351 - std::io::Write::write_fmt::haf55272405c09d9b
   3:     0x7fed21bf9642 - std::sys::backtrace::BacktraceLock::print::h61c3bd81a9458a03
   4:     0x7fed21bff709 - std::panicking::default_hook::{{closure}}::haf1ffb5d1e33a97f
   5:     0x7fed21bff233 - std::panicking::default_hook::hc32245deb6eaa988
   6:     0x7fed20bfd651 - std[836535f787e97d3]::panicking::update_hook::<alloc[bc4a0c534c311190]::boxed::Box<rustc_driver_impl[88352806430fedcc]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7fed21bffb2f - std::panicking::panic_with_hook::h43adc00fd0e494cb
   8:     0x7fed20c37211 - std[836535f787e97d3]::panicking::begin_panic::<rustc_errors[bfaf558bc9da2898]::ExplicitBug>::{closure#0}
   9:     0x7fed20c2b296 - std[836535f787e97d3]::sys::backtrace::__rust_end_short_backtrace::<std[836535f787e97d3]::panicking::begin_panic<rustc_errors[bfaf558bc9da2898]::ExplicitBug>::{closure#0}, !>
  10:     0x7fed20c2b097 - std[836535f787e97d3]::panicking::begin_panic::<rustc_errors[bfaf558bc9da2898]::ExplicitBug>
  11:     0x7fed20c42b01 - <rustc_errors[bfaf558bc9da2898]::diagnostic::BugAbort as rustc_errors[bfaf558bc9da2898]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7fed211de71a - rustc_middle[cc5e45d0ca739427]::util::bug::opt_span_bug_fmt::<rustc_span[4aa1f42928653b61]::span_encoding::Span>::{closure#0}
  13:     0x7fed211de9fa - rustc_middle[cc5e45d0ca739427]::ty::context::tls::with_opt::<rustc_middle[cc5e45d0ca739427]::util::bug::opt_span_bug_fmt<rustc_span[4aa1f42928653b61]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:     0x7fed211cf9bb - rustc_middle[cc5e45d0ca739427]::ty::context::tls::with_context_opt::<rustc_middle[cc5e45d0ca739427]::ty::context::tls::with_opt<rustc_middle[cc5e45d0ca739427]::util::bug::opt_span_bug_fmt<rustc_span[4aa1f42928653b61]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:     0x7fed1ef811c0 - rustc_middle[cc5e45d0ca739427]::util::bug::bug_fmt
  16:     0x7fed22b05015 - <rustc_trait_selection[64e2ca0a91c648aa]::traits::select::SelectionContext>::confirm_candidate
  17:     0x7fed22b1e68e - <rustc_trait_selection[64e2ca0a91c648aa]::traits::select::SelectionContext>::evaluate_trait_predicate_recursively
  18:     0x7fed22b299b5 - <rustc_trait_selection[64e2ca0a91c648aa]::traits::select::SelectionContext>::evaluate_root_obligation
  19:     0x7fed22b27c2f - rustc_traits[942bb15dcf444bf5]::evaluate_obligation::evaluate_obligation
  20:     0x7fed22b27569 - rustc_query_impl[f5b648d03701713a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f5b648d03701713a]::query_impl::evaluate_obligation::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 2usize]>>
  21:     0x7fed22b26748 - rustc_query_system[701c55d052347103]::query::plumbing::try_execute_query::<rustc_query_impl[f5b648d03701713a]::DynamicConfig<rustc_query_system[701c55d052347103]::query::caches::DefaultCache<rustc_type_ir[5960852b3f357f42]::canonical::CanonicalQueryInput<rustc_middle[cc5e45d0ca739427]::ty::context::TyCtxt, rustc_middle[cc5e45d0ca739427]::ty::ParamEnvAnd<rustc_middle[cc5e45d0ca739427]::ty::predicate::Predicate>>, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 2usize]>>, false, false, false>, rustc_query_impl[f5b648d03701713a]::plumbing::QueryCtxt, false>
  22:     0x7fed22b2639d - rustc_query_impl[f5b648d03701713a]::query_impl::evaluate_obligation::get_query_non_incr::__rust_end_short_backtrace
  23:     0x7fed22cf0473 - <rustc_trait_selection[64e2ca0a91c648aa]::traits::fulfill::FulfillProcessor as rustc_data_structures[99519894e5253d88]::obligation_forest::ObligationProcessor>::process_obligation
  24:     0x7fed22404e09 - <rustc_data_structures[99519894e5253d88]::obligation_forest::ObligationForest<rustc_trait_selection[64e2ca0a91c648aa]::traits::fulfill::PendingPredicateObligation>>::process_obligations::<rustc_trait_selection[64e2ca0a91c648aa]::traits::fulfill::FulfillProcessor>
  25:     0x7fed22e1c12a - <rustc_trait_selection[64e2ca0a91c648aa]::traits::fulfill::FulfillmentContext<rustc_trait_selection[64e2ca0a91c648aa]::traits::FulfillmentError> as rustc_infer[ac2fa1be7f77add4]::traits::engine::TraitEngine<rustc_trait_selection[64e2ca0a91c648aa]::traits::FulfillmentError>>::evaluate_obligations_error_on_ambiguity
  26:     0x7fed22e18a22 - rustc_hir_analysis[9991eadac4024967]::check::wfcheck::check_impl
  27:     0x7fed22c404c0 - rustc_hir_analysis[9991eadac4024967]::check::check::check_item_type
  28:     0x7fed22c3bfba - rustc_hir_analysis[9991eadac4024967]::check::wfcheck::check_well_formed
  29:     0x7fed22c3bf9b - rustc_query_impl[f5b648d03701713a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f5b648d03701713a]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 1usize]>>
  30:     0x7fed22c3b812 - rustc_query_system[701c55d052347103]::query::plumbing::try_execute_query::<rustc_query_impl[f5b648d03701713a]::DynamicConfig<rustc_data_structures[99519894e5253d88]::vec_cache::VecCache<rustc_span[4aa1f42928653b61]::def_id::LocalDefId, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[701c55d052347103]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f5b648d03701713a]::plumbing::QueryCtxt, false>
  31:     0x7fed22c3b316 - rustc_query_impl[f5b648d03701713a]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  32:     0x7fed22c389ef - rustc_hir_analysis[9991eadac4024967]::check::wfcheck::check_type_wf
  33:     0x7fed22c388fd - rustc_query_impl[f5b648d03701713a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f5b648d03701713a]::query_impl::check_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 1usize]>>
  34:     0x7fed2349f794 - rustc_query_system[701c55d052347103]::query::plumbing::try_execute_query::<rustc_query_impl[f5b648d03701713a]::DynamicConfig<rustc_query_system[701c55d052347103]::query::caches::SingleCache<rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[f5b648d03701713a]::plumbing::QueryCtxt, false>
  35:     0x7fed2349f576 - rustc_query_impl[f5b648d03701713a]::query_impl::check_type_wf::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7fed22fb2c3a - rustc_hir_analysis[9991eadac4024967]::check_crate
  37:     0x7fed22eb5bb0 - rustc_interface[3dfce13d5179a6f8]::passes::analysis
  38:     0x7fed22eb5877 - rustc_query_impl[f5b648d03701713a]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f5b648d03701713a]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 0usize]>>
  39:     0x7fed234a61e0 - rustc_query_system[701c55d052347103]::query::plumbing::try_execute_query::<rustc_query_impl[f5b648d03701713a]::DynamicConfig<rustc_query_system[701c55d052347103]::query::caches::SingleCache<rustc_middle[cc5e45d0ca739427]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[f5b648d03701713a]::plumbing::QueryCtxt, false>
  40:     0x7fed234a5e36 - rustc_query_impl[f5b648d03701713a]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7fed2373a6ca - <rustc_interface[3dfce13d5179a6f8]::passes::create_and_enter_global_ctxt<core[2e27404414be4892]::option::Option<rustc_interface[3dfce13d5179a6f8]::queries::Linker>, rustc_driver_impl[88352806430fedcc]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[2e27404414be4892]::ops::function::FnOnce<(&rustc_session[2fc1d960ad4ec558]::session::Session, rustc_middle[cc5e45d0ca739427]::ty::context::CurrentGcx, alloc[bc4a0c534c311190]::sync::Arc<rustc_data_structures[99519894e5253d88]::jobserver::Proxy>, &std[836535f787e97d3]::sync::once_lock::OnceLock<rustc_middle[cc5e45d0ca739427]::ty::context::GlobalCtxt>, &rustc_data_structures[99519894e5253d88]::sync::worker_local::WorkerLocal<rustc_middle[cc5e45d0ca739427]::arena::Arena>, &rustc_data_structures[99519894e5253d88]::sync::worker_local::WorkerLocal<rustc_hir[7611e5bcfcd2a5c8]::Arena>, rustc_driver_impl[88352806430fedcc]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  42:     0x7fed23594be2 - rustc_interface[3dfce13d5179a6f8]::interface::run_compiler::<(), rustc_driver_impl[88352806430fedcc]::run_compiler::{closure#0}>::{closure#1}
  43:     0x7fed2314a300 - std[836535f787e97d3]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[3dfce13d5179a6f8]::util::run_in_thread_with_globals<rustc_interface[3dfce13d5179a6f8]::util::run_in_thread_pool_with_globals<rustc_interface[3dfce13d5179a6f8]::interface::run_compiler<(), rustc_driver_impl[88352806430fedcc]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  44:     0x7fed23149fe4 - <<std[836535f787e97d3]::thread::Builder>::spawn_unchecked_<rustc_interface[3dfce13d5179a6f8]::util::run_in_thread_with_globals<rustc_interface[3dfce13d5179a6f8]::util::run_in_thread_pool_with_globals<rustc_interface[3dfce13d5179a6f8]::interface::run_compiler<(), rustc_driver_impl[88352806430fedcc]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[2e27404414be4892]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  45:     0x7fed2314896f - std::sys::thread::unix::Thread::new::thread_start::h4637f1bfded3ea68
  46:     0x7fed1cc94ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  47:     0x7fed1cd268c0 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81:0
  48:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.92.0 (ded5c06cf 2025-12-08) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [evaluate_obligation] evaluating trait selection obligation `Foo: HasComponent<()>`
#1 [check_well_formed] checking that `<impl at ../../try_pro/bug6.rs:6:1: 8:20>` is well-formed
... and 2 other queries... use `env RUST_BACKTRACE=1` to see the full query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0107`.
```

</p>
</details>
